### PR TITLE
=act #13866 fix ConcurrentModificationException in java.util.Map MDC

### PR DESCRIPTION
**Pull request by He-Pin**
fix this https://github.com/akka/akka/issues/13866 for release 2.3

### CODE REVIEW

**Comment by drewhk** [comment]
OK TO TEST

**Comment by hepin1989** [comment]
failed by SurviveNetworkInstabilitySpec

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 1]
```diff
@@ -1116,7 +1116,7 @@ trait DiagnosticLoggingAdapter extends LoggingAdapter {
    * These values can be used in PatternLayout when [[akka.event.slf4j.Slf4jLogger]] is configured.
    * Visit <a href="http://logback.qos.ch/manual/mdc.html">Logback Docs: MDC</a> for more information.
    */
-  def setMDC(jMdc: java.util.Map[String, Any]): Unit = mdc(if (jMdc != null) jMdc.asScala.toMap else emptyMDC)
+  def setMDC(jMdc: java.util.Map[String, Any]): Unit = mdc(if (jMdc != null) jMdc.asScala.foldLeft(Map.empty[String, Any])((map, kv) ⇒ map + kv) else emptyMDC)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: alternative: `.toSeq.toMap` ?

Comment (hepin1989) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: which one is better ,from the perspective of efficient?

Comment (patriknw) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: more efficient would probably be to skip the `toScala` and iterate `jMdc` entries and convert them to tuples (or use MapBuilder)

Comment (drewhk) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: maybe it is worth here to do the extra effort.

**Comment by drewhk** [comment]
Thanks for the contribution, LGTM!

**Comment by drewhk** [comment]
PLS BUILD

File: akka-actor/src/main/scala/akka/event/Logging.scala (hunk 1 above)
Comment (hepin1989) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: like this?
```scala
  def setMDC(jMdc: java.util.Map[String, Any]): Unit = {
    val mdcValue = if (jMdc != null) {
      val iterator = jMdc.entrySet().iterator()
      var builder = Map.newBuilder[String, Any]
      while (iterator.hasNext) {
        val entry = iterator.next()
        builder += (entry.getKey -> entry.getValue)
      }
      builder.result()
    } else emptyMDC
    mdc(mdcValue)
  }
```

Comment (drewhk) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: :+1:

Comment (patriknw) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: :+1:

Comment (hepin1989) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: OK,I will update then,thanks

File: akka-actor/src/main/scala/akka/event/Logging.scala [hunk 2]
```diff
@@ -1116,7 +1241,18 @@ trait DiagnosticLoggingAdapter extends LoggingAdapter {
...
-  def setMDC(jMdc: java.util.Map[String, Any]): Unit = mdc(if (jMdc != null) jMdc.asScala.toMap else emptyMDC)
+  def setMDC(jMdc: java.util.Map[String, Any]): Unit = {
+    val mdcValue = if (jMdc != null) {
+      val iterator = jMdc.entrySet().iterator()
+      val builder = Map.newBuilder[String, Any]
+      while (iterator.hasNext) {
+        val entry = iterator.next()
+        builder += (entry.getKey -> entry.getValue)
+      }
+      builder.result()
```
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: This is exactly what `jMdc.asScala.toMap` ends up doing, so I prefer the shorter version for readability.

**Comment by rkuhn** [comment]
Thanks for your work and energy, but I’m sorry to say that the ticket itself was faulty (as commented there).

File: akka-actor/src/main/scala/akka/event/Logging.scala (hunk 2 above)
Comment (hepin1989) [inline akka-actor/src/main/scala/akka/event/Logging.scala]: I miss the toMap call and think there is just asScala.
will close this.thanks

**Comment by hepin1989** [comment]
sorry for the misunderstanding .I am blind the toMap call

**Pull request closed without merging**